packages/babel-parser/src/parser/expression.js

```diff
@@ -1552,15 +1552,25 @@ export default class ExpressionParser extends LValParser {
   checkGetterSetterParams(method: N.ObjectMethod | N.ClassMethod): void {
     const paramCount = method.kind === "get" ? 0 : 1;
     const start = method.start;
-    if (method.params.length !== paramCount) {
+    // Allow `this` type annotation in TypeScript
+    if (
+      method.params.length === paramCount + 1 &&
+      method.params[0].type === "Identifier" &&
+      method.params[0].name === "this"
+    ) {
+      this.expectPlugin("typescript");
```

Comment:
When the first parameter is `this` the new implementation doesn't check the parameters count because the `} else if (method.params.length !== paramCount) {` branch is skipped:

```js
get prop(this: {}, foo, bar) {}
```

If we parsed `this` as a function parameter, then this plugin is already enabled (no need to assert). I would prefer to have as few as possible TypeScript-specific logic in the `parser/*` files. WDYT about extractions `method.kind === "get" ? 0 : 1` to a new `getAccessorsExpectedParamCount(method)` function and overwrite it in the typescript plugin?

Reply:
`method.params.length === paramCount + 1` wouldn't be true in that case I guess?

Reply:
Oh right :man_facepalming:

Reply:
@nicolo-ribaudo I have pushed a change for `getGetterSetterExpectedParamCount`. Thanks for suggesting the separation. I hadn't noticed I had the first usage of `this.expectPlugin("typescript")`.

At first, it felt a little odd that the error messages directly reference the required number of params despite that number being abstracted away. I considered extracting a `checkGetterSetterExpectedParamCount` method instead, but didn't find a good way to do so without duplicating logic in the TypeScript plugin. However, I found that there's less of a dissonance if I look at it as the error messages being based on target language and `getGetterSetterExpectedParamCount` being based on the source language, so I'm on board now 😄